### PR TITLE
P/jeffc/enable optimizations

### PR DESCRIPTION
**Pull request by Jeff-Ciesielski**

### CODE REVIEW

**Comment by Jeff-Ciesielski** [comment]
Resolves https://github.com/autosportlabs/ASL_F4_bootloader/issues/18 and https://github.com/autosportlabs/ASL_F4_bootloader/issues/12

File: Makefile [hunk 1]
```diff
@@ -79,8 +79,8 @@ LIBS = -lnosys
 
 LDFLAGS ?= --specs=nano.specs -lc -lgcc $(LIBS) -mcpu=$(CPU_TYPE) -g -gdwarf-2 
 LDFLAGS += -L. -Lcpu/common -L$(CPU_BASE) -T$(CPU_LINK_MEM) -Tlink_sections.ld 
-LDFLAGS += -nostartfiles -Wl,--gc-sections -mthumb -mcpu=$(CPU_TYPE)
-#LDFLAGS += -ffunction-sections -fdata-sections -flto -Os
+LDFLAGS += -nostartfiles -Wl,--gc-sections -mthumb -mcpu=$(CPU_TYPE) -Os
```
Comment (stieg) [inline Makefile:82]: So that is a C optimization flag. I guess GCC passes that optimization to the linker somehow?

Comment (Jeff-Ciesielski) [inline Makefile:82]: It does (surprisingly).  Apparently starting with 4.8, additional optimizations can be performed at link time.

File: cpu/stm32f3xx/cdc_acm/usb_pwr.c [hunk 2]
```diff
@@ -37,7 +37,7 @@
 /* Private macro -------------------------------------------------------------*/
 /* Private variables ---------------------------------------------------------*/
 __IO uint32_t bDeviceState = UNCONNECTED; /* USB device status */
-__IO bool fSuspendEnabled = FALSE;  /* true when suspend is possible */
+__IO bool fSuspendEnabled = TRUE;  /* true when suspend is possible */
```
Comment (stieg) [inline cpu/stm32f3xx/cdc_acm/usb_pwr.c:40]: What is the benefit in doing this?

File: img_utils.c [hunk 3]
```diff
@@ -114,6 +111,11 @@ void jump_to_app(uint32_t address)
 	__set_MSP((uint32_t)(vec_table[0]));
 	__set_PSP((uint32_t)(vec_table[0]));
 
+	/* Perform a pipeline flush to ensure stack pointers are set
+	 * before jumping */
```
Comment (stieg) [inline img_utils.c]: Nit... comment formatting.

``` c
/*
 * I am a
 * multiline comment
 */
```

File: cpu/stm32f3xx/cdc_acm/usb_pwr.c (hunk 2 above)
Comment (Jeff-Ciesielski) [inline cpu/stm32f3xx/cdc_acm/usb_pwr.c:40]: Ensuring that the drivers used on the app and bootloader are configured in exactly the same manner.  

For 1.0.0, I'd like to do a review of the USB stack in it's entirety.

**Comment by stieg** [comment]
Minor nits.  Fix as appropriate and merge.

File: cpu/stm32f3xx/cdc_acm/usb_pwr.c (hunk 2 above)
Comment (stieg) [inline cpu/stm32f3xx/cdc_acm/usb_pwr.c:40]: Sure.

File: Makefile (hunk 1 above)
Comment (stieg) [inline Makefile:82]: So since we are on 4.7 what does this do?

Comment (stieg) [inline Makefile:82]: Oh... so this just does code opt... no link opt.  Interestingly enough the linker says that its O value must be a positive int (like `-O1`) and that the int value (other than being positive) doesn't matter.

**Pull request merged**